Updates permission descriptions files comparer and refactors project

irvinesunday opened the pull request:
Fixes https://github.com/microsoftgraph/msgraph-permissions-scraper/issues/16

This PR:
- Updates the logic used to check whether there was a diff. between permissions descriptions fetched from the Service Principal vs GitHub. Rather than just checking equality based on the serialized string values of the aforementioned permissions descriptions, we now evaluate equality based on a given scope description `id` value and the contents contained in the descriptions specified under that `id`. 
- Refactors common functionalities out of the `DescriptionsScraper` TimerTrigger Function class into a common module. This is to help in separation of concerns and aid in testability.
- Writes tests against the above refactored-out common module. Adds sample test JSON files.
- Adds utility functions to help with common functionalities - parameter checking.
- Refactors some methods to help with conciseness, readability and maintainability.

irvinesunday committed:
Initial code refactor to make permissions conversion modular

irvinesunday committed:
Use latest language ver.

irvinesunday committed:
Add method to update permissions; code refactor; add XML comments

irvinesunday committed:
Add tests and test files

irvinesunday committed:
Utility function for null checks

irvinesunday committed:
Update helper method to be more clear in its functionality

irvinesunday committed:
Update assembly to allow testing of internal classes

irvinesunday committed:
Refactor out methods for separation of concerns and testability

irvinesunday committed:
Add string extensions class

irvinesunday committed:
Code refactoring

irvinesunday committed:
Merge remote-tracking branch 'origin/dev' into enhancement/is/change-comparer-function

irvinesunday committed:
Comment updates

baywet commented on PermissionsScraper/PermissionsScraper/Common/StringExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+        /// <param name="rawString">The raw input string</param>
+        /// <param name="newLine">The new line break.</param>
+        /// <returns>The changed string.</returns>
+        public static string ChangeLineBreaks(this string rawString, string newLine = "\n")
```
maybe https://docs.microsoft.com/en-us/dotnet/api/system.environment.newline?view=net-5.0  ?

irvinesunday replied:
Are you suggesting I use `Environment.NewLine` instead of `\n` as the default value?
The default value needs to be a compile-time constant.

baywet commented on PermissionsScraper/PermissionsScraper/PermissionsScraper.csproj:
```diff
@@ -1,7 +1,8 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <TargetFramework>netcoreapp3.1</TargetFramework>
```
bump to net5.0 if possible to remove errors

baywet commented on PermissionsScraper/PermissionsScraper/Helpers/ProtectedApiCallHelper.cs:
```diff
@@ -23,16 +24,10 @@ internal static class ProtectedApiCallHelper
-            HttpClient httpClient = new HttpClient();
+            HttpClient httpClient = new();
             HttpRequestHeaders defaultRequestHeaders = httpClient.DefaultRequestHeaders;
             if (!defaultRequestHeaders.Accept.Any(m => m.MediaType == "application/json"))
```
```suggestion
            if (!defaultRequestHeaders.Accept.Any(m => "application/json".Equals(m.MediaType, StringComparison.OrdinalIgnoreCase)))
```

irvinesunday replied:
Why not `if (!defaultRequestHeaders.Accept.Any(m => m.MediaType.Equals("application/json", StringComparison.OrdinalIgnoreCase)))` ?

This is easier to read in my opinion.

baywet replied:
because the style I'm suggesting protects you against MediaType being null, the one you're suggesting doesn't

baywet commented on PermissionsScraper/PermissionsScraper/Services/PermissionsProcessor.cs:
```diff
@@ -0,0 +1,149 @@
+                var permissionsDescriptions = permissionsDescriptionsToken?.Value<JArray>(scopeName)?.ToObject<List<Dictionary<string, object>>>();
+                if (permissionsDescriptions == null) continue;
+
+                if (referencePermissionsDescriptions.ContainsKey(scopeName) is false)
```
```suggestion
                if (!referencePermissionsDescriptions.ContainsKey(scopeName))
```

baywet commented on PermissionsScraper/PermissionsScraper/Services/PermissionsProcessor.cs:
```diff
@@ -0,0 +1,149 @@
+                {
+                    var id = permissionDescription["id"];
+                    var permissionExists = referencePermissionsDescriptions[scopeName].Exists(x => x.ContainsValue(id));
+                    if (permissionExists is false)
```
```suggestion
                    if (!permissionExists)
```

baywet commented on PermissionsScraper/PermissionsScraper/Services/PermissionsProcessor.cs:
```diff
@@ -0,0 +1,149 @@
+            UtilityFunctions.CheckArgumentNull(permissionsDescriptionsToken, nameof(permissionsDescriptionsToken));
+            UtilityFunctions.CheckArgumentNull(referencePermissionsDescriptions, nameof(referencePermissionsDescriptions));
+
+            foreach (string scopeName in scopesNames)
```
all of that could be replaced by a few linq statements...

irvinesunday replied:
In this PR [comment](https://github.com/microsoftgraph/microsoft-graph-devx-api/pull/636#discussion_r669058046) you did mention that semantically, `foreach` would signify that we want to perform an _action_ for each of the element in the `IEnumerable`, whereas a `linq` `Select` signifies we're _converting_ each element of an the `IEnumerable` to a different thing. In our case we intend to perform several _actions_ using our element.

baywet replied:
from reading the code it looks to me that these loops are projecting two collections, retaining elements of the projection based on conditions. In my experience linq is really good at dataset manipulation (projections, correlations, filtering,...) and is easier to read than loops for that (besides the potential performance gains). I do reserve loops for actions, and by actions I mean calling methods on subobjects/classes.

irvinesunday replied:
I've tried replacing the `foreach` with `Linq`, however the `ref` params trigger the error message: "_Cannot use ref, out, or in parameter inside an anonymous method, lambda expression, query expression or local function._"

baywet commented on PermissionsScraper/PermissionsScraper/Services/PermissionsProcessor.cs:
```diff
@@ -0,0 +1,149 @@
+                        var referencePermissionsText = JsonConvert.SerializeObject(referencePermission, Formatting.Indented);
+                        var updatablePermissionText = JsonConvert.SerializeObject(updatablePermission, Formatting.Indented);
+
+                        if (referencePermissionsText.Equals(updatablePermissionText, StringComparison.OrdinalIgnoreCase) is not true)
```
```suggestion
                        if (!referencePermissionsText.Equals(updatablePermissionText, StringComparison.OrdinalIgnoreCase))
```

baywet reviewed (changes requested):
lots of opportunity to get rid of convoluted loops here...

irvinesunday committed:
Use net5.0

irvinesunday committed:
System.Linq with net5.0 currently throws an error

'System.Linq, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

irvinesunday committed:
Use net5.0 in pipeline build

baywet commented on PermissionsScraper/PermissionsScraper/Triggers/DescriptionsScraper.cs:
```diff
@@ -67,7 +66,7 @@ public static void Run([TimerTrigger("%ScheduleTriggerTime%")] TimerInfo myTimer
                 }
 
-                if (!_servicePrincipalPermissions.Any())
+                if (_servicePrincipalPermissions.Count == 0)
```
why remove any?

irvinesunday replied:
I was getting this error when I bumped up the project to use net5.0 
`'System.Linq, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.`

baywet replied:
`C:\Users\runneradmin\.nuget\packages\microsoft.net.sdk.functions\3.0.3\build\Microsoft.NET.Sdk.Functions.Build.targets(41,5): error : It was not possible to find any compatible framework version [D:\a\msgraph-permissions-scraper\msgraph-permissions-scraper\PermissionsScraper\PermissionsScraper\PermissionsScraper.csproj]` leads me to think the azure functions SDK on the agents might be outdated. Maybe adding a reference to 3.0.13 would fix this issue.
https://www.nuget.org/packages/Microsoft.NET.Sdk.Functions/3.0.13

baywet replied:
additional resources https://docs.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-process-guide
https://techcommunity.microsoft.com/t5/apps-on-azure/net-on-azure-functions-roadmap/ba-p/2197916

baywet replied:
https://github.com/microsoftgraph/msgraph-permissions-scraper/pull/18/files#diff-f13d431b675a71faf2e36d49c905263a140f64a4ae96f96d6c183aad3c6059de

irvinesunday replied:
Tried bumping up to this version: https://www.nuget.org/packages/Microsoft.NET.Sdk.Functions/3.0.13 but still error-ing out. After reading this article you shared: https://techcommunity.microsoft.com/t5/apps-on-azure/net-on-azure-functions-roadmap/ba-p/2197916, I'm of the opinion we can just stick to `netcoreapp3.1` in the interim until we are sure of full support for `net5.0` for the .NET isolated process in Azure Functions.

irvinesunday committed:
Use quotation marks around version

irvinesunday committed:
Revert to using net core 3.1

irvinesunday committed:
Revert to use LINQ

irvinesunday committed:
Use default C# language

This is because the default language shipped with the specified dotnet version is used in the build pipeline.

irvinesunday committed:
Use C# 8.0 language features

irvinesunday committed:
Use C# 8.0 features in the Resources.cs class

irvinesunday committed:
Update PermissionsScraper/PermissionsScraper/Services/PermissionsProcessor.cs

Co-authored-by: Vincent Biret <vibiret@microsoft.com>

irvinesunday committed:
Update PermissionsScraper/PermissionsScraper/Services/PermissionsProcessor.cs

Co-authored-by: Vincent Biret <vibiret@microsoft.com>

irvinesunday committed:
Update PermissionsScraper/PermissionsScraper/Helpers/ProtectedApiCallHelper.cs

Co-authored-by: Vincent Biret <vibiret@microsoft.com>

irvinesunday committed:
Replace foreach

andrueastman commented on PermissionsScraper/PermissionsScraper.Tests/Services/PermissionsProcessorTests.cs:
```diff
@@ -0,0 +1,113 @@
+                        Dictionary<string, List<Dictionary<string, object>>> gitHubPermissionsDict)
+                        GetPermissionsDescriptionsDictionaries()
+        {
+            var servicePrincipalPermissionsText = GetServicePrincipalPermissionsDescriptionsText().GetAwaiter().GetResult();
```
Can we make the function async?

irvinesunday committed:
Change method signatures to async

andrueastman reviewed (approved):
👍🏼 On the tests.
